RevenueCat/Plugins/iOS/RCCommonFunctionality.m

```diff
@@ -181,31 +224,63 @@ + (void)makeDeferredPurchase:(RCDeferredPromotionalPurchaseBlock)deferredPurchas
             completion(nil, [self payloadForError:error withExtraPayload:@{@"userCancelled": @(userCancelled)}]);
         } else {
             completion(@{
-                @"purchaserInfo": purchaserInfo.dictionary,
-                @"productIdentifier": transaction.payment.productIdentifier
+                    @"purchaserInfo": purchaserInfo.dictionary,
+                    @"productIdentifier": transaction.payment.productIdentifier
             }, nil);
         }
     });
 }
 
 + (void)checkTrialOrIntroductoryPriceEligibility:(nonnull NSArray<NSString *> *)productIdentifiers
-                                 completionBlock:(void (^)(NSDictionary<NSString *, NSDictionary *> *))completion
+                                 completionBlock:(RCReceiveIntroEligibilityBlock)completion
 {
     NSAssert(RCPurchases.sharedPurchases, @"You must call setup first.");
-    
+
     [RCPurchases.sharedPurchases checkTrialOrIntroductoryPriceEligibility:productIdentifiers completionBlock:^(NSDictionary<NSString *,RCIntroEligibility *> * _Nonnull dictionary) {
         NSMutableDictionary *response = [NSMutableDictionary new];
         for (NSString *productID in dictionary) {
             RCIntroEligibility *eligibility = dictionary[productID];
             response[productID] = @{
-                @"status": @(eligibility.status),
-                @"description": eligibility.description
+                    @"status": @(eligibility.status),
+                    @"description": eligibility.description
             };
         }
         completion([NSDictionary dictionaryWithDictionary:response]);
     }];
 }
 
++ (void)paymentDiscountForProductIdentifier:(NSString *)productIdentifier
+                                   discount:(nullable NSString *)discountIdentifier
+                            completionBlock:(RCHybridResponseBlock)completion
+{
+    if (@available(iOS 12.2, *)) {
+        NSAssert(RCPurchases.sharedPurchases, @"You must call setup first.");
+        [self productWithIdentifier:productIdentifier completionBlock:^(SKProduct *_Nullable aProduct) {
+            if (aProduct) {
+                SKProductDiscount *discountToUse = [self discountWithIdentifier:discountIdentifier forProduct:aProduct];
+
+                if (discountToUse) {
+                    [RCPurchases.sharedPurchases paymentDiscountForProductDiscount:discountToUse product:aProduct completion:^(SKPaymentDiscount *paymentDiscount, NSError *error) {
+                        if (paymentDiscount) {
+                            self.discounts[[paymentDiscount.timestamp stringValue]] = paymentDiscount;
+                            completion(paymentDiscount.dictionary, nil);
+                        } else {
+                            completion(nil, [self payloadForError:error  withExtraPayload:@{}]);
+                        }
+                    }];
+                } else {
```

Comment:
[not needed for this PR] I kinda feel like these are complex enough that they might merit being in their own class. 

`RCCommonFunctionality` is right now acting as both a Facade and implementing logic for products & discounts. The latter feels a bit out of scope.